Network changes for k_timeout_t

jukkar opened the pull request:
Initial set of patches for networking related code. This converts most of the networking APIs to use k_timeout_t and should be merged first.

jukkar committed:
net: dns: Refactor because of timeout overhaul

Use k_timeout_t internally, no change to user API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: sntp: Refactor because of timeout overhaul

Use 64-bit time in order to avoid overlaps, and do not use K_MSEC()
as that will convert to k_timeout_t which we do not want in this
case.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: http: Refactor because of timeout overhaul

Use k_timeout_t internally, no change to user API.
Clarify the documentation of the timeout parameter that it is
in milliseconds.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: websocket: Refactor because of timeout overhaul

Mention in websocket API documentation that the timeout value
is in milliseconds. Check timeout values properly using K_TIMEOUT_EQ()
macro.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: sockets: Refactor because of timeout overhaul

Use k_timeout_t internally, no change to user API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: gptp: Refactor because of timeout overhaul

Use k_timeout_t internally, no change to user API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: lldp: Refactor because of timeout overhaul

Use k_timeout_t internally, no change to user API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ppp: Refactor because of timeout overhaul

Use k_timeout_t internally, no change to user API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: context: Refactor because of timeout overhaul

The net_context API will change, the s32_t timeout parameter
will be changed to k_timeout_t. All the Zephyr users of this API will
be changed in subsequent commits. This is internal Zephyr API only,
so the API is not deprecated etc.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tcp: Refactor because of timeout overhaul

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ipv6: nbr: Refactor because of timeout overhaul

Convert minimal parts to support k_timeout_t values.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tcp2: Refactor because of timeout overhaul

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: trickle: Refactor because of timeout overhaul

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: mgmt: Refactor because of timeout overhaul

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ipv4: autoconf: Refactor because of timeout overhaul

Convert minimal parts to support k_timeout_t values.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: promisc: Refactor because of timeout overhaul

Convert minimal parts to support k_timeout_t values.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: if: Refactor because of timeout overhaul

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: dhcpv4: Refactor because of timeout overhaul

Convert minimal parts to support k_timeout_t values.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: pkt: Refactor because of timeout overhaul

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: echo_server: Fix compile issues from timeout overhaul

Feed proper time delay parameter to K_THREAD_DEFINE()

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: arp: Refactor because of timeout overhaul

Convert minimal parts to support k_timeout_t values.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Convert k_sleep() to k_msleep()

Needed because of k_timeout_t changes.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: config: Refactor because of timeout overhaul

Convert minimal parts to support k_timeout_t values.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: conn_mgr: Fix time delay in K_THREAD_DEFINE()

The time parameter cannot be K_NO_DELAY because the time delay
parameter is ms so using value 0 instead.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: echo_client: Fix compile issues from timeout overhaul

Use proper timeout value for poll()

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: net_mgmt: Fix compile issues from timeout overhaul

Feed proper time delay parameter to K_THREAD_DEFINE()

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jhedberg commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -1161,15 +1177,17 @@ int z_impl_zsock_poll(struct zsock_pollfd *fds, int nfds, int timeout)
 				break;
 			}
 
-			if (timeout == K_NO_WAIT) {
+			if (K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
 				break;
 			}
 
-			if (timeout != K_FOREVER) {
-				/* Recalculate the timeout value. */
-				remaining_time = time_left(entry_time, timeout);
-				if (remaining_time <= 0) {
+			if (K_TIMEOUT_EQ(timeout, K_FOREVER)) {
```
Looks like a missing `!` here since the original comparison was `timeout != K_FOREVER`

jukkar replied:
Yep, fixed

jukkar committed:
net: ieee802154: Fix the scan duration data type

Converting to k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: coap: Fix the timeout data type

Converting timeout to k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: getaddrinfo: Fix timeout data type

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: buf: Remove timeout variable from debug print

The timeout value cannot be printed anymore after k_timeout_t
changes.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: udp: Fix timeout value type

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
tests: net: route: Fix timeout data type

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: dumb_http_server_mt: Fix K_THREAD_DEFINE() call

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
shell: telnet: Refactor because of timeout overhaul

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: syslog: Fix to use proper data type for sleep

Needed because of k_timeout_t conversion

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: big_http_download: Fix to use proper data type for sleep

Needed because of k_timeout_t conversion

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: zperf: Refactor because of timeout overhaul

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Converted most of the networking code to use k_timeout_t.
Still missing:
* lwm2m
* mqtt
* canbus related stuff
* modem
I am hoping volunteers for converting above.

rlubos commented:
@jukkar I can handle MQTT.

pfalcon commented:
@jukkar, so does this have to go together with another PR, as in: if another PR is merged, and this isn't, we have broken build? If so, can you link to this PR in the description?

And if not, the usual comment applies: can we please avoid gigantic, hard to review PRs? Can this be split into a few smaller PRs, e.g. per-sub-subsystem to allow for attentive review? (And actually, could have been done gradually, like convert one part of code, see it working well (or break), use that experience to convert next part of code).

jukkar commented:
> so does this have to go together with another PR, as in: if another PR is merged, and this isn't, we have broken build? If so, can you link to this PR in the description?

This PR should go in first as we are not disabling legacy timeout api yet.

> 
> And if not, the usual comment applies: can we please avoid gigantic, hard to review PRs? Can this be split into a few smaller PRs, e.g. per-sub-subsystem to allow for attentive review? (And actually, could have been done gradually, like convert one part of code, see it working well (or break), use that experience to convert next part of code).

There are lot of commits but they are small and should be easy to review.
I am not going to start to split this one, the amount of commits would not be lower anyway.
The sanity passes so I am hoping I did not make a mistake, but of course more testing is needed after this is merged.

rlubos reviewed (commented):
I have not spotted anything serious, just some nits.

jukkar committed:
net: Add NET_WAIT_FOREVER symbol to be used as timeout in network APIs

If a network API expects a millisecond timeout, then NET_WAIT_FOREVER
symbol can be used to indicate that the timeout should last forever.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented on include/net/net_pkt.h:
```diff
@@ -1119,12 +1119,12 @@ void net_pkt_print_frags(struct net_pkt *pkt);
  *        If K_NO_WAIT, then return immediately. If K_FOREVER, then
  *        wait as long as necessary. Otherwise, wait up to the specified
- *        number of milliseconds before timing out.
+ *        number of time before timing out.
```
Yes, you version is much better.

jukkar commented:
Updated according to comments. Removed K_MS_TIMEOUT and introduced NET_WAIT_FOREVER to be used in those networking APIs that use millisecond timeouts.

jukkar commented on samples/net/sockets/coap_client/src/coap-client.c:
```diff
@@ -42,7 +42,7 @@ static struct coap_block_context blk_ctx;
 static void wait(void)
 {
-	if (poll(fds, nfds, K_FOREVER) < 0) {
+	if (poll(fds, nfds, K_MS_FOREVER) < 0) {
```
I used NET_WAIT_FOREVER symbol in the latest version of this PR but that can be changed if we decide that plain -1 is to be used.

stephanosio replied:
With `poll` being a standard POSIX function and the standard clearly stating that the timeout value of `-1` means "forever," shouldn't it be perfectly okay to use `-1` here?

> If the value of timeout is -1, poll() shall block until a requested event occurs or until the call is interrupted.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/poll.html

jukkar replied:
Agreed. The value of NET_WAIT_FOREVER is also -1 so that could be used too.

pfalcon replied:
> I used NET_WAIT_FOREVER symbol in the latest version of this PR but that can be changed if we decide that plain -1 is to be used.

Looking at https://github.com/zephyrproject-rtos/zephyr/pull/24149/files#diff-e986daadfca28716ec435fc4c722ae40L28 , that indeed appears to be the plan. Using different ways to specify the same behavior for different APIs in Zephyr would be rather confusing. Maybe @pabigot as the author of #24149 can comment here.

Otherwise, as @stephanosio points out, as long as values used are consistent with "well-known APIs" (so, the values themselves are well-known), I think it's fine to skip symbolic constants and use these well-known, intuitive (enough) values directly. After all, we don't write `TWO + TWO`.

@jukkar, Also compare our discussion on a similar matter https://github.com/zephyrproject-rtos/zephyr/pull/24124#pullrequestreview-388278757 . Of all other matters, consistency is important. That's why I used "-1" in a sample config to mean "forever"/"infinite", and let 0 mean 0 (skipping not pertinent discussion whether 0 makes sense or not in a particular case or nor - if it doesn't, users just won't use it there). And surely, we can have different representations in different places - if there're good reasons to do so. If not, consistency should prevail.

pabigot replied:
@pfalcon Regarding #24149, originally the thread start delay parameter in `K_THREAD_DEFINE` was specified in milliseconds.  At the time we started moving from integral millisecond timeouts to `k_timeout_t` (last fall), the API changed to take a `k_timeout_t` value for consistency with `k_thread_create()`, so 0 was converted to `K_NO_WAIT` and -1 to `K_FOREVER`.  When the delay changed back to milliseconds (this spring) forever was converted back to -1, which may be just a canonical representation for an unlimited wait specified by a negative value (I haven't looked).

The `K_THREAD_DEFINE` API does not actually document support for "wait-forever" (never start it automatically) regardless of that being useful, and continues to recommend `K_NO_WAIT` for no-delay even though that is now wrong as it's a timeout value.

rlubos reviewed (approved):
LGTM

andyross commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1367,10 +1381,15 @@ pkt_alloc_with_buffer(struct k_mem_slab *slab,
+		if (remaining <= 0) {
+			timeout = K_NO_WAIT;
+		} else {
+			timeout = Z_TIMEOUT_TICKS(remaining);
```
FWIW: "K_NO_WAIT" is just a timeout of zero, and the Z_TIMEOUT_*() APIs include the clamp for you.  You can just use Z_TIMEOUT_TICKS(end - z_tick_get()) if you like.

andyross reviewed (approved):
One note on a simplification (which applies in a few places, I only put the remark in one spot)

pfalcon commented on subsys/net/ip/net_shell.c:
```diff
@@ -1645,7 +1645,7 @@ static int cmd_net_dns_query(const struct shell *shell, size_t argc,
 
 #if defined(CONFIG_DNS_RESOLVER)
-#define DNS_TIMEOUT K_MSEC(2000) /* ms */
+#define DNS_TIMEOUT (MSEC_PER_SEC * 2) /* ms */
```
Check the trailing comment please.

jukkar replied:
What is there to check? The DNS_TIMEOUT is in milliseconds.

pfalcon replied:
> What is there to check? The DNS_TIMEOUT is in milliseconds.

Hmm, ok. I would guess that `/* ms */` comment is a left-over from old time when there was a raw number, like `2000 /* ms */`. Thus, I hinted at removing it, as in other places where there're unit-providing macros, we don't use it. But yes, formally, it's still in ms.

pfalcon commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -34,7 +34,7 @@ LOG_MODULE_REGISTER(net_dns_resolve, CONFIG_DNS_RESOLVER_LOG_LEVEL);
 #define LLMNR_IPV6_ADDR "[ff02::1:3]:5355"
 
-#define DNS_BUF_TIMEOUT 500 /* ms */
+#define DNS_BUF_TIMEOUT K_MSEC(500) /* ms */
```
I have to admit I didn't follow the timeout refactor (which would be 3rd or 4th in my count, or part of grande plan of which I already lost track). So in the very these commit, I see the changes like:

> #define DNS_TIMEOUT K_SECONDS(2)
> #define DNS_TIMEOUT (2 * MSEC_PER_SEC)

And then here:

> #define DNS_BUF_TIMEOUT 500 /* ms */
#define DNS_BUF_TIMEOUT K_MSEC(500) /* ms */

So, we replace `K_SECONDS()` with `* MSEC_PER_SEC` and yet introduce more uses of `K_MSEC()`?

I fail to understand nature and purpose of such changes. It all just goes in vicious circles, without clear purpose or improvement. My suggestions: a) if K_SECONDS() was removed, call to reinstate it; if it was deprecated, call to undeprecate it; b) skip doing purposeless changes like this, this will shrink this PR of following diffstats: +501 −376 to much more modest numbers. (Any line of changed lines may contain a typo, which we'll catch in a year due to lack of thorough testing).

jukkar replied:
The timeouts in this example are used for different APIs. The buf timeout is needed for buffer allocation where the timeout type is k_timeout_t. The dns timeout is for timeout where the type is s32_t.

pfalcon reviewed (commented):
I find many changes introduce superfluous, not sure if it's a problem of this PR or the "upstream" timeout refactor.

jukkar commented:
Updated according to comments.
